### ISSUE TITLE
SSL approach?

### ISSUE THREAD

**Issue by pfalcon** [micropython/micropython#952]
So, I've been prototyping uPy-specific package manager. Did tar (not tar.gz) unpacking, then decided to look into downloading of dependencies. And here I noticed that PyPI forces all access to SSL now. Kaboom. All is lost. Unless we bring SSL in. Any ideas?

**Comment by silpol** [comment]
you may reconsider SSL/TLS with [mbed SSL](https://tls.mbed.org/source-code) (former PolarSSL)
* their license is [GPL plus FOSS exception](https://tls.mbed.org/how-to-get);
* their memory/code footprint was reasonable when I checked it last time;
* they had greater exposure (in terms of installations) and time-wise;
* they had been [acquired by ARM](https://tls.mbed.org/tech-updates/blog/polarssl-part-of-arm), so it will be embedded for good.

**Comment by MattMatic** [comment]
I've been using XySSL > PolarSSL > mbed SSL for some years now in a commercial embedded product. We did try with AxTLS and that was fine, but the PolarSSL library has moved much faster.

Currently PolarSSL is GPL or commercial, but earlier versions were BSD licensed too. Once fully migrated, mbed-SSL will be Apache-style licensed.

I can definitely help and assist with PolarSSL/mbedSSL. (In fact, I wrote & contributed the SSH/SFTP libraries using the PolarSSL math libraries - hopefully they'll pop out into the main distribution :)

Be aware though that TLS implementation requires a hunk of RAM. Crypto core parts may be ok (e.g. RSA, AES, SHA-256 etc), but taken together the TLS implementation needs a lot of working space. PolarSSL does work very well indeed and is nicely abstracted from the sockets layer (so you could apply TLS over some other transport) ;)
Matt

**Comment by pfalcon** [comment]
It's indeed outrages fact that ARM pushes mbed SSL, licensed on GPL, onto unsuspecting customers, they're together on the brink, or already violating GPL, so would be good target for http://gpl-violations.org/ (if it were alive). That's not the first time ARM screws up doing open-source things, e.g. previous case: http://dangerousprototypes.com/2013/05/03/mbeds-bsd-b-s/ . Their way going forward is indeed only to relicense it under more liberal open-source license, but that's got to be hard if PolarSSL was a real open-source project were many people contributed (they need to ask re-licensing permission from everyone).

Summing up, we (disclaimer: in the worst case I, a project contributor) will avoid using GPL software in MicroPython. That's because we respect both GPL and our users, including those who don't know that GPL compliance (especially in business settings) require good deal of planning, good will, and overall effort. We won't be putting them in liability position by making them use something they don't understand and ready to comply with.

**Comment by pfalcon** [comment]
Otherwise, on the technical topic, I kinda lost interest so far in implementing SSL in Python. That's because that ESP8266 for example already uses axTLS, so if making proper port for it, axTLS support in uPy will come "automagically".

**Comment by danicampora** [comment]
Now that you mention that about the ESP8266, the CC3200 also has its own
flavor of TLS/SSL along with SHA,MD5,DES,AES engines built-in in the
hardware.

**Comment by MattMatic** [comment]
@pfalcon: outrages notwithstanding, my understanding is that mbed SSL will be Apache style (which is similar to MIT & BSD), not GPL. It is true that PolarSSL _was_ GPL + commercial but as I commented, it used to be BSD + commercial.

**Comment by silpol** [comment]
When it comes to hardware built-in engines, I'd rather to be cautious. Personal past experience with TI OMAP chips at work shows - their built-in engines had been slower than software implemented ones without any optimization. Also we had some attempts to offload encry/decry to clean cpu badnwidth for other apps, but the gain had been rather mediocre. Maybe modern built-ins could be better - it requires verification. Oh, and one another thing - one can not fix hardware in case of security advisory, only re-configure software away from built-ins.

**Comment by pfalcon** [comment]
@MattMatic: Yes, I certainly appreciate the info, I heard about mbed SSL, but didn't look further than GPL license stamp, if there will be more specific info regarding re-licensing later, I'd appreciate a ping here.

**Comment by MattMatic** [comment]
@pfalcon: it's worth looking for the XySSL BSD licensed version to get a feel for the library. There were some architectural changes migrating to PolarSSL, but the basic flavour of the two are the same. As soon as ARM have switched to Apache for mbed SSL I'll try and remember to ping here ;)
I found it one of the easiest SSL/TLS libraries to work with, lots of optimisation options, and very good support from Paul Bakker (though support is now community based since the acquisition).

**Comment by pfalcon** [comment]
> It's indeed outraged fact that

As expected, ARM was caught on this one too:

http://www.cnx-software.com/2015/04/30/arm-microcontrollers-are-better-because-theres-no-gpl/

**Comment by pfalcon** [comment]
I started prototyping SSL module.

@dpgeorge : How do we handle axtls dependency?

**Comment by pfalcon** [comment]
I as usual hacked on it, and not sure how upstream-mergeable the changes. Anyway, what about having a git submodule?

**Comment by pfalcon** [comment]
Also, what about bigint implementations - both uPy and axtls has own, so code will be duplicated. Any ideas?

**Comment by pfalcon** [comment]
Ok, 25K raw code size is the minimum client size I can get with axtls.

**Comment by dpgeorge** [comment]
> 25K raw code size is the minimum

Does that include shared bignum code, or duplicated?

> Anyway, what about having a git submodule?

How big is the code (num files, wc -l)?  Do we want to start incorporating submodules?

**Comment by pfalcon** [comment]
> Does that include shared bignum code, or duplicated?

Duplicated. axtls bignum code is 3.2K though (at the minimum size settings).

>How big is the code (num files, wc -l)? 

~ 20 files spread around at least 3 dirs, >250K size.

>Do we want to start incorporating submodules?

That's what I'm asking ;-). For this case, IMHO it makes sense.

**Comment by dpgeorge** [comment]
Yes, ok, let's go for a submodule for axTLS.  Put it somewhere sensible in lib/ directory.  The framework for using SSL should be modular so we (or others) can switch out axTLS for something else if needed.

How to handle the case of modifying the axTLS code?

**Comment by pfalcon** [comment]
> Yes, ok, let's go for a submodule for axTLS. 

Thanks.

> Put it somewhere sensible in lib/ directory.

Not extmod/ ? IIRC we have a convention - if only extmod needs a dependency, it goes into extmod/. Ah, right, somehow you guessed I don't have a module in extmod/ so far. But if written like described below, that's the most natural location for it.

> The framework for using SSL should be modular so we (or others) can switch out axTLS for something else if needed.

Well, SSL is layered thing, how it works is that you pass it a socket, it returns an object which wraps it with the same interface (read()/write()). My initial prototype deals with OS fd sockets, but it of course should be rewritten to take an arbitrary stream object and use that. So, an SSL module will automagically work with any underlying socket implementation (or not socket implementation, for example SSL over SPI should be possible). Of course, that requires axTLS support (or patching it). And per above, any sane SSL implementation would easily plug in (proving that is left as an exercise to people who need it though).

> How to handle the case of modifying the axTLS code?

We discuss that with @projectgus now, hopefully we will be able to come to a common repo (or my repo being fork of his). Hopefully we also will motivate each other come talk to upstream sooner rather than later. But otherwise, just the same as handling changes to re1.5/uzlib (process to push updates to uPy repo is a bit different of course).

**Comment by pfalcon** [comment]
>> Put it somewhere sensible in lib/ directory.
> Not extmod/?

Well, with https://github.com/micropython/micropython/commit/a9058bf2944db20f5e2377a6a38c37a46d3d6aa0 ,  I agree that external libs rather go to lib/ dir for consistency.

**Comment by pfalcon** [comment]
Lately, mbedTLS was relicensed under Apache license: https://tls.mbed.org/tech-updates/releases/mbedtls-2.1.0-released-under-apache . That doesn't change previous plan to use axTLS, but gives another, final excuse to not try to clean up my previous patchset, and thus treat it as is instead, because further plan may be as well to transition to mbedTLS instead. So, I pushed whatever I had to https://github.com/pfalcon/axtls , what's left is to add it as a submodule to uPy and integrate its building into "deplibs" target of unix Makefile. upip already works with ussl module.

**Comment by pfalcon** [comment]
axTLS support is fully merged and available for any port.

**Comment by xyb** [comment]
Thank you, @pfalcon. But aaa8867 not include AES? Do we have a plan for that?
